transformer_engine/pytorch/module/linear.py

```diff
@@ -1340,7 +1340,7 @@ def forward(
             if get_ub(self.ub_name + "_dgrad").is_fp8_ubuf():
                 fp8_grad = True
 
-        with self.prepare_forward(
+        with torch.cuda.device(inp.device), self.prepare_forward(
```

Comment:
If we go this route, it seems better to use the weight's device. Presumably that's where all the FP8 state and other GPU data has been initialized (and if not, then there are bigger problems).
```suggestion
        with torch.cuda.device(getattr(self, self.weight_names[0])), self.prepare_forward(
```

Reply:
I changed to the weight device for linear-based layers. DPA still uses `query_layer.device` because it doesn't have any weights.